### ISSUE TITLE
Abnormally high ram usage on qtile

### ISSUE THREAD

**Issue by trianglecirclexyz** [qtile/qtile#2484]
This is my first issue posted on GitHub, sorry if I forget to include important information/improperly format. Also apologies if this is a duplicate of another issue, I tried to make sure it isn't to the best of my abilities before posting.

# Issue description

Qtile's ram usage starts increasing right after boot up. I've left it on for a few hours and it has gotten as high as 6 gigabytes. System restarts don't fix the issue resets ram usage to normal levels, but it continues growing right after. Is this a possible memory leak?
![image](https://user-images.githubusercontent.com/25913782/119397217-a71aeb00-bc8a-11eb-96d1-7ec5a1af64be.png)

# Qtile version

0.17.0

# Stack traces

[qtile.log](https://github.com/qtile/qtile/files/6534605/qtile.log)
[xsession-errors.txt](https://github.com/qtile/qtile/files/6534607/xsession-errors.txt)

**Comment by m-col** [comment]
Hi @ryynison please could you post your config?

When you restart Qtile, does the memory usage drop down and slowly increase over time, or is it always high?

**Comment by ryynison** [comment]
Hey @m-col, thanks for getting back to me. 

The memory usage starts at 50ish megabytes but increases over time as you said. It goes back to 50 after i restart qtile using super+ctrl+R, but continues to grow again after. In about 4 hours, it reaches around 2 gigabytes.
[config.py.txt](https://github.com/qtile/qtile/files/6580750/config.py.txt)

**Comment by ramnes** [comment]
Anything interesting out of `qtile top`?

**Comment by ryynison** [comment]
```
[raiyan@reno-arch ~]$ qtile top
tracemalloc not started on qtile, start by setting PYTHONTRACEMALLOC=1 before starting qtile 
or force start tracemalloc now, but you'll lose early traces
```

**Comment by m-col** [comment]
:D yes... could you set PYTHONTRACEMALLOC=1 before starting Qtile and then maybe
check on `qtile top` every now and then and see how it progresses? It will
hopefully give us an insight into where the memory is being used

On 210601 05:53, Raiyan wrote:
>```
***@***.*** ~]$ qtile top

**Comment by ramnes** [comment]
> ```
> [raiyan@reno-arch ~]$ qtile top
> tracemalloc not started on qtile, start by setting PYTHONTRACEMALLOC=1 before starting qtile 
> or force start tracemalloc now, but you'll lose early traces
> ```

Dude... It literally tells you what to do, please do not spam my notification inbox with such messages.

**Comment by ryynison** [comment]
Yeah my apologies, I was being an idiot. Here is the proper output for `qtile top`:

```
Qtile - Top 10 lines
#   Line                                     Memory
1   <frozen importlib._bootstrap_external>:647 9609.6 KiB

2   libqtile/window.py:1116                  986.6 KiB
    arr = array.array("B", data)
3   python3.9/linecache.py:137               757.9 KiB
    lines = fp.readlines()
4   python3.9/abc.py:85                      362.7 KiB
    cls = super().__new__(mcls, name, bases, namespace, **kwargs)
5   <frozen importlib._bootstrap_external>:1560 304.8 KiB

6   x11/xcbq.py:67                           251.4 KiB
    r.setdefault(v, []).append(k)
7   <frozen importlib._bootstrap_external>:1568 170.7 KiB

8   site-packages/pyparsing.py:3958          153.2 KiB
    self.errmsg = "Expected " + _ustr(self)
9   collections/__init__.py:497              122.5 KiB
    result = type(typename, (tuple,), class_namespace)
10  site-packages/pyparsing.py:4091          112.5 KiB
    self.strRepr = "{" + " ".join(_ustr(e) for e in self.exprs) + "}"

14350 other: 8559.9 KiB
Total allocated size: 21392.0 KiB
```

**Comment by m-col** [comment]
Doesn't look too bad yet, could you update when you're up to really high memory usage? Thanks. 2 GB is crazy!

**Comment by ryynison** [comment]
This is the output for `qtile top` after a few hours (qtile memory grew to around 2 GB).:

```
Qtile - Top 10 lines
#   Line                                     Memory
1   dbus/service.py:661                      365866.5 KiB
    args = message.get_args_list(**parent_method._dbus_get_args_options)
2   <frozen importlib._bootstrap_external>:647 9609.7 KiB

3   libqtile/window.py:1116                  1004.0 KiB
    arr = array.array("B", data)
4   python3.9/linecache.py:137               757.9 KiB
    lines = fp.readlines()
5   python3.9/abc.py:85                      362.6 KiB
    cls = super().__new__(mcls, name, bases, namespace, **kwargs)
6   <frozen importlib._bootstrap_external>:1560 304.8 KiB

7   x11/xcbq.py:67                           251.4 KiB
    r.setdefault(v, []).append(k)
8   <frozen importlib._bootstrap_external>:1568 170.7 KiB

9   site-packages/pyparsing.py:3958          153.2 KiB
    self.errmsg = "Expected " + _ustr(self)
10  collections/__init__.py:497              122.5 KiB
    result = type(typename, (tuple,), class_namespace)

14459 other: 9001.8 KiB
Total allocated size: 387605.2 KiB
```

**Comment by trianglecirclexyz** [comment]
Fixed in 0.18
